src/main/python/rlbot/matchcomms/client.py

```diff
@@ -66,22 +69,31 @@ def close(self):
         assert not self.thread.is_alive()
 
 
-async def read_into_queue(websocket: WebSocketClientProtocol, incoming: Queue):
+async def read_into_queue(websocket: WebSocketClientProtocol, incoming: Queue, team):
     async for message in websocket:
         try:
-            incoming.put(json.loads(message))
+            decoded_message = json.loads(message)
         except json.decoder.JSONDecodeError:
             print('Failed to decode this message:', repr(message))
             traceback.print_exc()
+        # make sure we can't see the message if it is just for the other team
+        if ('team' in decoded_message and 'team_only' in decoded_message and
+                team is not None and decoded_message['team_only'] == True and
+                decoded_message['team'] != team):
+            continue
+        incoming.put(decoded_message)
 
-async def send_from_queue(websocket: WebSocketClientProtocol, outgoing: Queue):
+async def send_from_queue(websocket: WebSocketClientProtocol, outgoing: Queue, team):
     logger = get_logger('matchcomms_json_encode')
     while True:
         # TODO: use something like https://github.com/aio-libs/janus to avoid polling
         while outgoing.empty():
             await asyncio.sleep(0.01)
 
         obj = outgoing.get_nowait()
+        # if we aren't a script, add our team to the packet so that team-only matchcomms can work
+        if team is not None:
+            obj["team"] = team
```

Comment:
Seems like this makes the 'team' attribute reserved. If somebody already had team in their message, e.g. something like
```
{ name: 'MyBot', team: ['MyAlly1', 'MyAlly2']}
```
then this would overwrite their intended data, right? That might be OK, but we should think of ways to mitigate that problem. Is there a different implementation that could avoid it? If not, are people OK with it? Could we at least warn people if their data is getting overwritten?

Reply:
good idea. maybe something like:
```python
if obj.get("team"): print("Overriding the 'team' key in your match communication object!")
obj["team"] = team
```

Reply:
Another angle on it: why does this belong in the matchcomms layer rather than the TMCP layer?

Reply:
> why does this belong in the matchcomms layer rather than the TMCP layer?
Because the team of the sender has to be known so the client knows who it came from and if they should be allowed to see it

Reply:
I like Virx's idea.
Maybe: 
```python
if "team" in obj and obj["team"] != team:
    print("Overriding the 'team' key in your match communication object!")
obj["team"] = team
```

Reply:
The discourage sending packets as another bot (out of sight, out of mind)

Reply:
This is why I wanted two matchcomms servers, one per team. It would also restrict the spying and malicious messages.

Reply:
I actually thought I responded to this PR to this effect but I guess my connection was so bad it failed.

Reply:
The bot can just bypass the client and send the message to whatever server it wants - just like a bot can send inputs for any other bot.

Reply:
> This is why I wanted two matchcomms servers, one per team. It would also restrict the spying and malicious messages.

Two things: One, how do we prevent people from joining (maliciously or accidentally) the wrong server? Two, this seems much more breaking for scripts etc. than simply reserving `team` and telling people if it is being changed. 

I'm not sure there's enough public, un-update-able bots using matchcomms yet for this edge case to have happened, and warning people if it does happen so they can change it seems like enough to me.

Reply:
Scripts would have two matchcomm clients.

Each client is locked to their server. Afaik, there isn't a way to access another teams server, since it is stored in the botmanager instance which bots don't have access to.

Reply:
> Each client is locked to their server. Afaik, there isn't a way to access another teams server, since it is stored in the botmanager instance which bots don't have access to.

Assuming by "since it is stored in the botmanager" you mean "since each server is stored in the botmanager," *how* would each client be locked to its server? What prevents someone from just directly accessing the servers with custom clients? The server is just that - a server. Who can access it is not restricted, and I'm not sure it can be.

Reply:
If they don't have the port they can't make a custom client

Reply:
How do you restrict it so that the server and the authorized clients know the port but not the unauthorized ones?

Reply:
> > why does this belong in the matchcomms layer rather than the TMCP layer?
> 
> Because the team of the sender has to be known so the client knows who it came from and if they should be allowed to see it

I don't understand the reasoning. Isn't there a TMCP client in python that could be initialized with the proper team?

Reply:
Using fewer pronouns for clarity.

Passing the team in the packet isn't a TMCP problem. (Well, team-only matchcomms was made in part because TMCP wants it, but that's not what this is about.) The problem here is making sure potential receivers of the packet know what team the sender is on so the potential receivers know if they are allowed to look at it. The only way that I have thought of to pass along the team information is by putting it in the packet, since the packet is the only thing that's shared at all.

Reply:
I know that you're intending to give the team only functionality to more than just TMCP, I'm questioning whether that's in line with the design philosophy of matchcomms. As I understand it, matchcomms was designed by Dom to be totally unbiased about how it gets used. Generally that's a smart way to avoid regrets later, e.g. if some idea comes along that you didn't imagine at first, and now the system is too specific to be useful and gets replaced.

This is a one-way door, this change cannot be undone. I think team filtering is broadly useful, but I don't feel good about risking the longevity of matchcomms. I recommend either:
- Getting approval from @DomNomNom or,
- Making this a third layer that sits between TCMP and matchcomms.

Reply:
> I recommend either:
> - Getting approval from DomNomNom or,
> - Making this a third layer that sits between TCMP and matchcomms.

As a third party layer we already have my handler (tmcp package) which filters by team, and also specifies team in messages.

Reply:
The team-only is an extra feature, not mandatory. Setting `team_only` to `True` in the packet will prevent the packet from being sent to the other team. Otherwise, it will be sent to both teams. (At least this was the original idea)

Reply:
To complete the middle layer, you'd need to make a utility that helps you send standard team-stamped (and index-stamped?) messages which is separate from the core matchcomms implementation. Otherwise you're still narrowing matchcomms in an irreversible manner.